### ISSUE TITLE
Some bugs & enhancement suggestions

### ISSUE THREAD

**Issue by ChiwaCime** [railmapgen/rmg#399]
## TL;DR:
1. Station list table would float above side panels on mobile UA.
2. Custom color picker seems to call the native color picker, and in some case scenario (like most Android Browser), is nearly unusable. 
3. Branch spacing is not enough to meet spacings for an 5+-line-intechange station in GZMTR style.
5. Weird path would shown when rail map padding is over 50%.
6. Closed branch path is missing (not transparent or hidden) when the current station is on the main line in Shanghai Metro style.
7. Muitl-branch system is not well-supported in MTR style & Shanghai Metro style.
8. Version of the RMG on GitLab page is outdated (v3.12.10).

## Bug 1:
The station list would float above side panels and covers parts of the panels, making it unable to do adjustment to the rail map. 
![Screenshot 2022-07-13 150635](https://user-images.githubusercontent.com/51258794/178683162-65689876-79d7-4df4-b276-9eb3a1029cad.png)
(Mozilla/5.0 (iPhone; CPU iPhone OS 14_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Mobile/15E148 Safari/604.1)
![Screenshot_20220712-222946_Kiwi Browser](https://user-images.githubusercontent.com/51258794/178683203-ac5599b6-3e88-47e7-9d89-7d5fe73931a7.png)
(Kiwi Browser 98.0.4758.57)
![Screenshot_20220712-223101_Firefox](https://user-images.githubusercontent.com/51258794/178683217-cbe9c77e-2d1e-426d-9e99-8567af82fef0.png)
(Firefox stable 102.2.0 #2015890451)
Theoretically this problem is general, but it only occurs when using an browser with portrait UA, mostly smartphones' browser, where width is not enough.
Temporarily solution including using desktop UA, max out canvas scale and flip smartphones to landscape.
> 1.（bug）部分安卓端在V5版本UI内容显示不全以至于无法关闭部分编辑页面
> 1. (Bug) the UI content of some Android terminals in V5 is not fully displayed, so that some editing pages cannot be closed

我们会保证UI在发布时会在移动设备上正确显示，当然项目仍然是以desktop为目标来开发的。
We will ensure the correctness of UI in the public release, but do keep in mind this project is made for the desktop environment.

_Originally posted by @thekingofcity in https://github.com/railmapgen/rmg/issues/276#issuecomment-1066084377_
## Enhancement 2:
Custom color picker seems to call the native color picker in v5 rather than using the built-in one back in v3. This could be a nice improvement for desktop users.
However, in some case scenario (like most Android Browser), is nearly unusable. Here're examples.
![Screenshot_20220713-120600_Firefox](https://user-images.githubusercontent.com/51258794/178683315-ae4a1242-b8cc-4512-a91f-e4ee6b601cc9.png)
(Firefox stable 102.2.0 #2015890451)
Could it be possible to re-integrade the hex color input?
![Screenshot_20220713-120542_Firefox](https://user-images.githubusercontent.com/51258794/178683373-a5988d7a-88df-4d8a-9e4d-f1fb3da7783f.png)
## Enhancement 3:
Branch spacing is not enough. 
An 5+-line-intechange station on upper branch line would cover the below station name with 3 characters at the main line in GZMTR style.
![Screenshot 2022-07-13 142320](https://user-images.githubusercontent.com/51258794/178683432-903633b3-685a-472c-8c56-e8b09a1a96af.png)
2-line-intechange station for an express stopping station.
![Screenshot 2022-07-13 133838](https://user-images.githubusercontent.com/51258794/178683442-89f8beaf-d5d1-4a00-bf3a-b150f9cb91d0.png)
## Enhancement 4：
![Screenshot_20220713-124824_Firefox](https://user-images.githubusercontent.com/51258794/178683504-f0c3fa0c-2868-4764-8696-a2df556d4643.png)
As shown.
## Bug 5:
![Screenshot 2022-07-13 133252](https://user-images.githubusercontent.com/51258794/178683562-384107ad-177e-4191-84a1-63e2b6d20524.png)
As shown.
> 无法显示支线

这里基本可以确认你当时设支线的两端设错了。上海地铁风格的线路支线只有在支线两端都接入主线且当前车站不在支线上的时候会发生支线线路不显示的bug（如下图3和图4，这里用的预设模版中的港铁东铁线，注意当前车站在主支线上不同车站的区别）。（不过这个也是V3时期就有的老bug了，希望@thekingofcity 还是尽早修复一下吧）
![rmg 7gy7 火炭 Fo_Tan](https://user-images.githubusercontent.com/98044618/166111762-1b575ef5-5f9e-4e00-821e-37310063c1f6.png)
![rmg hybz 馬場 Racecourse](https://user-images.githubusercontent.com/98044618/166111769-af361925-c29b-4513-b80e-cc262f2825c3.png)

_Originally posted by @52PD in https://github.com/railmapgen/rmg/issues/336#issuecomment-1114005956_
### Bug 6:
This Bug has been appeared since the initial release (v1.0).
Here's an example for indoor map in Shanghai Metro style. The end of branches cannot lineup properly.
![Screenshot 2022-07-13 133554](https://user-images.githubusercontent.com/51258794/178683614-50907261-65d8-4ca3-b588-07e908c94576.png)
To line up, the indoor width would have to be at least 8000px.
![Screenshot 2022-07-13 144732](https://user-images.githubusercontent.com/51258794/178683638-f97198f0-3320-4fd4-b0a6-e91a2cc05e11.png)
As for the MTR style, the number will be only 3500.
![Screenshot 2022-07-13 154158](https://user-images.githubusercontent.com/51258794/178683656-f0f7a5ae-d722-4f41-98df-694a3cd06cb9.png)
![Screenshot 2022-07-13 154320](https://user-images.githubusercontent.com/51258794/178683967-4baea3ac-bc10-4bb7-b5fd-0a702462b687.png)
### Enhancement 7:
![Screenshot 2022-07-13 142620](https://user-images.githubusercontent.com/51258794/178688258-0273992b-8e59-46d4-80c9-9013bd945d8e.png)
As shown.

**Comment by 52PD** [comment]
再补充一个：注意在Safari和Chrome上生成的广州地铁站名图的区别（上为Safari，下为Chrome）
![rmg l1mz 番禺广场 Panyu_Square-2](https://user-images.githubusercontent.com/98044618/178793385-8e3bd254-e35b-4e5b-9c41-e901843ab631.png)
![rmg l1mz 番禺广场 Panyu_Square (1)](https://user-images.githubusercontent.com/98044618/178793396-5b79d8a3-9762-4ef6-9810-ae9de230abf2.png)

**Comment by wongchito** [comment]
1. Pinned column of station grid stacks on top of side panel
    - This will be fixed in [rmg-components](https://github.com/railmapgen/rmg-components)
    - z-index style needs to be added to rmg-side-panel
2. Advanced colour picker
    - This is not planned at the moment. We expect this to be implemented by each UA itself. 
    - Safari iOS version is an example. 
3. More branch spacing (GZMTR)
    - Branch spacing is now set as absolute value. We can consider change it to percentage (as against to SVG height). 
    - However this change may breaks the existing templates. We need to find out a non-breaking way to fix this.
4. Bugs when padding > 50% (GZMTR)
    - We will limit the max padding percentage. 
5. Leave to @thekingofcity 
6. Diverging branch path calculation
    1. MTR - This is planned as a follow up fix to #365 to fix all types of branches.
    2. SHMetro
7. GitLab and other instances
    - Track in #356 
8. English station name alignment inconsistent across browsers
    - Track in #400

**Comment by 52PD** [comment]
新的<del>特性</del>bug：请在图中找出“上传配置文件”的按钮😂
![Screen Shot 2022-07-15 at 8 53 45 PM](https://user-images.githubusercontent.com/98044618/179327998-189349d1-8c0f-4332-8fcf-fe6a3e73839f.png)

（以下是明说部分）好像最近更新后改车站名修改模版参数什么的窗口开始挡点了“上传”和“下载”后的下拉菜单了😂

**Comment by wongchito** [comment]
> 新的~特性~bug：请在图中找出“上传配置文件”的按钮😂 ![Screen Shot 2022-07-15 at 8 53 45 PM](https://user-images.githubusercontent.com/98044618/179327998-189349d1-8c0f-4332-8fcf-fe6a3e73839f.png)
> 
> （以下是明说部分）好像最近更新后改车站名修改模版参数什么的窗口开始挡点了“上传”和“下载”后的下拉菜单了😂



> 新的~特性~bug：请在图中找出“上传配置文件”的按钮😂 ![Screen Shot 2022-07-15 at 8 53 45 PM](https://user-images.githubusercontent.com/98044618/179327998-189349d1-8c0f-4332-8fcf-fe6a3e73839f.png)
> 
> （以下是明说部分）好像最近更新后改车站名修改模版参数什么的窗口开始挡点了“上传”和“下载”后的下拉菜单了😂

Fixed in railmapgen/rmg-components@d5e76e6

**Comment by 52PD** [comment]
> Fixed in [railmapgen/rmg-components@d5e76e6](https://github.com/railmapgen/rmg-components/commit/d5e76e6) and [1404f42](https://github.com/railmapgen/rmg/commit/1404f42b06b65dd5797aa02c96c9d44bc10ca460)

辛苦了👍

**Comment by thekingofcity** [comment]
嗨，感谢您对项目的支持与建议。现在来说明一下bug 5。两端闭合的支线是明确在上海风中不被支持的功能，原因是上海地铁目前无此类线路，没有可以参考的实现形式，感谢您的理解。

> 嗨，感谢您汇报的问题。不过当前上海风格暂不支持闭合的下方支线，即支线的左右两端都接入主线的情况。您无法在上海风格中创建此类型的线路。您可以尝试闭合的上方支线，但我亦无法确认此种情况是否被支持。如果您需要上海风格中环线的样式，可以参考上海地铁4号线的模板，谢谢。
> 
> _Originally posted by @thekingofcity in https://github.com/railmapgen/rmg/issues/377#issuecomment-1140394305_

另，关于bug 2颜色选择器在安卓上几乎不可用的问题，我感觉留一个input允许用户直接输入hex作为临时解决方案会比较好一些？ @wongchito

**Comment by 52PD** [comment]
> 嗨，感谢您对项目的支持与建议。现在来说明一下bug 5。两端闭合的支线是明确在上海风中不被支持的功能，原因是上海地铁目前无此类线路，没有可以参考的实现形式，感谢您的理解。

emm。。。这样的话要不要考虑在上海风格中彻底隐藏掉这种两端闭合的支线（包括车站）部分啊（就像外部路线那种，在没设定外部路线线路名和颜色的时候直接隐藏），毕竟像现在这种情况下出现一堆浮空车站真的挺怪的🤔

而且这个浮空车站bug最神奇的地方就是：如果当前车站选的正好是这种两端闭合支线的车站上的时候支线线路又会显示出来（见我发的那个上海风的东铁线的（当前车站为马场的那个）图），这样的话真的很难让别的使用者相信“上海风不支持两端闭合支线”

**Comment by thekingofcity** [comment]
> emm。。。这样的话要不要考虑在上海风格中彻底隐藏掉这种两端闭合的支线（包括车站）部分啊（就像外部路线那种，在没设定外部路线线路名和颜色的时候直接隐藏），毕竟像现在这种情况下出现一堆浮空车站真的挺怪的🤔
> 
> 而且这个浮空车站bug最神奇的地方就是：如果当前车站选的正好是这种两端闭合支线的车站上的时候支线线路又会显示出来（见我发的那个上海风的东铁线的（当前车站为马场的那个）图），这样的话真的很难让别的使用者相信“上海风不支持两端闭合支线”

如果你是说创建的时候隐藏，那已经隐藏了。如果你说是已有这种支线存在，在画布上隐藏的话，我不认为是个值得修改的改进。因为1. 用户切换时会默认所有元素都存在，消失的而非错误的显示反而会让用户迷惑，毕竟我们没有任何其他地方展示不兼容的部分，不如fail fast instead of fail safe。而且rmg在设计之初就是考虑了多风格切换的可能性。2. 很难判定两端闭合的支线，至少以目前我的认知来看。进行这样的修正会耗费过多的精力而无法带来可见的用户体验提升。

**Comment by rainydew** [comment]
说起来，上海地铁风格使用固定的站名换乘线间距和中英文同一行以后，破坏了之前的数据
现在间距可以批量应用在所有车站
是否中英文同一行的配置也设置在所有车站？
分开来设置比较麻烦，得下载了json用脚本批量修改

**Comment by thekingofcity** [comment]
> 说起来，上海地铁风格使用固定的站名换乘线间距和中英文同一行以后，破坏了之前的数据
> 现在间距可以批量应用在所有车站
> 是否中英文同一行的配置也设置在所有车站？
> 分开来设置比较麻烦，得下载了json用脚本批量修改

感谢建议，默认自动转换为在同一行没有考虑到已有存档的样式兼容性。目前已推送修改后的更新脚本到线上，车站在升级存档时中英文默认不在同一行。

另：使用以下脚本在浏览器的开发者工具中可以快速更改同一行设置，所有车站的`one_line`属性都将被置为`false`，在脚本执行完毕后刷新页面即可看到修改后的结果。

```javascript
const param = JSON.parse(localStorage.getItem('rmgParam'));
const stn_list = Object.fromEntries(Object.entries(param.stn_list).map(([id, stnInfo]) => {
  stnInfo.one_line=false;
  return [id, stnInfo]；
}));
param.stn_list = stn_list;
localStorage.setItem('rmgParam', JSON.stringify(param));
```
